### PR TITLE
Add LogitTrackingProcessor

### PR DESCRIPTION
**Pull request by cpfiffer**
This PR adds `LogitTrackingProcessor`, a logit processor that wraps around any other processor to store unstructured and structured logits through the sampled sequence. I needed this code elsewhere and figured it is popular enough to upstream into `outlines`.

I have included documentation on the processors, which doesn't exist currently. Tests are included as well.

`LogitTrackingProcessor` makes it easy to perform analysis on disagreements between structured and unstructured tokens. It will be of benefit to researchers, educators, and users who wish to debug their Outlines generators.

An example plot for a regex requiring four digits. This is the distribution of token probabilities on the first token.

image

Using the tracker is simple:

```python
from outlines import generate, models
from outlines.processors import add_tracking
from pydantic import BaseModel
import pandas as pd

model = models.transformers("HuggingFaceTB/SmolLM2-135M-Instruct")
tokenizer = model.tokenizer.tokenizer

class Person(BaseModel):
    name: str
    age: int

# Create generator with tracking
generator = generate.json(model, Person)

# Convenience wrapper to add tracking
generator = add_tracking(generator)

# Apply templating
prompt = tokenizer.apply_chat_template(
    [{"role": "system", "content": "You are a helpful assistant, responding in JSON."}, {"role": "user", "content": "Make me a person with a name, age, zip code, and state. Return the JSON only."}],
    tokenize=False,
    add_bos=True,
    add_generation_prompt=True,
)

# Generate the response
generator(prompt)

# Retrieve the top-k tokens
top_k = generator.logits_processor.get_top_tokens(k=5)

# Get unstructured logits
for position_dict in top_k:
    position_dict['position'] # 0,1,2, etc
    position_dict['text_so_far'] # Text at this point in the sequence

    for token in position_dict['tokens']:
        token['token'] # The token
        token['unstructured_prob'] # Probability of the token in the unstructured distribution
        token['structured_prob'] # Probability of the token in the structured distribution
        token['unstructured_logit'] # Logit of the token in the unstructured distribution
        token['structured_logit'] # Logit of the token in the structured distribution
        token['is_chosen'] # Whether the token was actually sampled

# Convert to dataframe
df = generator.logits_processor.to_dataframe(show="probs", min_value=0.01)
#    position token   natural  constrained  chosen
# 0         0   You  0.021324          0.0   False
# 1         0   The  0.021959          0.0   False
# 2         0  Sure  0.025492          0.0   False
# 3         0  JSON  0.031045          0.0   False
# 4         0    To  0.031047          0.0   False

# Get the token sequence up to position 5
generator.logits_processor.sequence(5)
```

**NOTE:** Currently, the tracking processor does not support batch processing. I recommend deferring this to a later PR.

Related: 

- https://github.com/dottxt-ai/outlines/issues/479 for the primary discussion of having access to the underlying logits
- https://github.com/dottxt-ai/outlines/issues/614
- https://github.com/dottxt-ai/outlines/issues/1055 as this can be used as a debugging tool

### CODE REVIEW

**Comment by cpfiffer** [comment]
There's a few `mypy`-related tests to resolve but I think the meat of this is ready for review.

Current issues:
```python
outlines/processors/tracking.py:22: error: Library stubs not installed for "pandas"  [import-untyped]
outlines/processors/tracking.py:22: note: Hint: "python3 -m pip install pandas-stubs"
outlines/processors/tracking.py:22: note: (or run "mypy --install-types" to install all missing stub packages)
outlines/processors/tracking.py:22: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
outlines/processors/tracking.py:105: error: Item "list[Any]" of "Any | Any | list[Any] | Any" has no attribute "shape"  [union-attr]
outlines/processors/tracking.py:108: error: Item "list[Any]" of "Any | Any | list[Any] | Any" has no attribute "shape"  [union-attr]
outlines/processors/tracking.py:275: error: Value of type "dict[str, list[Any] | Any] | None" is not indexable  [index]
outlines/processors/tracking.py:276: error: Value of type "dict[str, list[Any] | Any] | None" is not indexable  [index]
outlines/processors/tracking.py:508: error: "SequenceGenerator" has no attribute "logits_processor"  [attr-defined]
outlines/processors/tracking.py:512: error: "SequenceGenerator" has no attribute "logits_processor"  [attr-defined]
outlines/processors/tracking.py:515: error: "SequenceGenerator" has no attribute "logits_processor"  [attr-defined]
outlines/processors/tracking.py:516: error: "LogitTrackingProcessor" has no attribute "tokenizer"  [attr-defined]
outlines/processors/tracking.py:516: error: "SequenceGenerator" has no attribute "logits_processor"  [attr-defined]
outlines/processors/tracking.py:519: error: "SequenceGenerator" has no attribute "logits_processor"  [attr-defined]
```

File: outlines/processors/tracking.py [hunk 1]
```diff
@@ -0,0 +1,521 @@
...
+
+        # Get tokens up to the specified position
+        end_pos = len(self.chosen_tokens) if pos is None else pos
+        tokens_to_decode = self.chosen_tokens[:end_pos]
+
+        # Decode the sequence
+        return "".join(tokenizer.decode(tokens_to_decode))
+
+
+def add_tracking(generator: "SequenceGenerator") -> "SequenceGenerator":
```
Comment (rlouf) [inline outlines/processors/tracking.py:476]: I would name this `track_logits` instead.

File: docs/reference/processors.md [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+Users can track the token probabilities and logits at each step of the generation process using the `LogitTrackingProcessor`. This is useful for debugging and understanding the generation process.
+
+## Available Processors
+
+Outlines provides several specialized processors for different use cases:
+
+- `JSONLogitsProcessor`: Ensures generation follows a JSON schema
+- `RegexLogitsProcessor`: Constrains generation to match a regex pattern
+- `CFGLogitsProcessor`: Enforces a context-free grammar
+- `LogitTrackingProcessor`: Analyzes token probabilities during generation
```
Comment (rlouf) [inline docs/reference/processors.md:14]: Tracks/Accumulates/Logs token probabilities ?

File: docs/reference/processors.md [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+- `GuideLogitsProcessor`: Base class for guided generation
```
Comment (rlouf) [inline docs/reference/processors.md:15]: Do we need to document the base class?

File: docs/reference/processors.md [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+### Important notes about logit tracking
+
+- Tracking logits is a slow operation, so do not use it in production environments
+- The processor will accumulate logits if you call `generator(prompt)` multiple times, meaning that the tokens stored can be aggregated across generations. You can use `generator.logits_processor.clear()` to reset the processor, or construct a new generator and call `add_tracking` again to start tracking from scratch.
+- Processed logits will contain `-inf` values when structured outputs are used
+- Token decoding requires the wrapped processor to have a tokenizer attribute
+- Memory usage grows linearly with sequence length
+- The tracking processor only supports single-batch processing
+- Tracking logits can incur significant overhead -- do not use it in production environments
```
Comment (rlouf) [inline docs/reference/processors.md:196]: I would add an example of how we can use these logits processors directly with e.g. transformer pipes

File: outlines/processors/tracking.py [hunk 5]
```diff
@@ -0,0 +1,521 @@
...
+import torch
+from numpy.typing import NDArray
+
+from .base_logits_processor import OutlinesLogitsProcessor, Array
+
+if TYPE_CHECKING:
+    from outlines.generate import SequenceGenerator
+
+# Try importing pandas, but don't fail if not available
+try:
```
Comment (rlouf) [inline outlines/processors/tracking.py:21]: Trying to import `pandas` in the methods that require it instead is cleaner imo

File: outlines/processors/tracking.py [hunk 6]
```diff
@@ -0,0 +1,521 @@
...
+        unstructured_probs = [
+            torch.softmax(torch.tensor(logits), dim=-1).numpy()
+            for logits in self.unstructured_logits
+        ]
+        structured_probs = [
+            torch.softmax(torch.tensor(logits), dim=-1).numpy()
+            for logits in self.structured_logits
+        ]
+
+        if as_matrix:
```
Comment (rlouf) [inline outlines/processors/tracking.py:170]: Do we need the `as_matrix` parameter? It seems like a straightforward transformation that users can do themselves should they need it

Comment (rlouf) [reply outlines/processors/tracking.py:170]: I think we need to make a choice here, what's the downside of returning a NumPy array every time?

**Review by rlouf — APPROVED** [review]

**Comment by rlouf** [comment]
Looks good to me, just a few comments

**Review by rlouf — CHANGES_REQUESTED** [review]
Great addition! I left a few comments that need to be addressed before merging.

File: outlines/processors/tracking.py (hunk 6 above)
Comment (cpfiffer) [reply outlines/processors/tracking.py:170]: If we choose one, I think probably using the `as_matrix=True` is the best. I'll update it, thanks!

**Comment by cpfiffer** [comment]
I've addressed the comments, appreciated!

I have some remaining code style issues that are kind of out of my knowledge at the moment, if people have tips on these I'd love them. Otherwise I'll have to come back to this in a week or two.

```python
outlines/processors/tracking.py:242: error: No overload variant of "__getitem__" of "list" matches argument type "tuple[slice[None, None, None], int]"  [call-overload]
outlines/processors/tracking.py:242: note: Possible overload variants:
outlines/processors/tracking.py:242: note:     def __getitem__(self, SupportsIndex, /) -> Any
outlines/processors/tracking.py:242: note:     def __getitem__(self, slice[Any, Any, Any], /) -> list[Any]
outlines/processors/tracking.py:243: error: No overload variant of "__getitem__" of "list" matches argument type "tuple[slice[None, None, None], int]"  [call-overload]
outlines/processors/tracking.py:243: note: Possible overload variants:
outlines/processors/tracking.py:243: note:     def __getitem__(self, SupportsIndex, /) -> Any
outlines/processors/tracking.py:243: note:     def __getitem__(self, slice[Any, Any, Any], /) -> list[Any]
outlines/processors/tracking.py:246: error: Value of type "dict[str, Any | Any | list[Any] | Any] | None" is not indexable  [index]
outlines/processors/tracking.py:246: error: No overload variant of "__getitem__" of "list" matches argument type "tuple[slice[None, None, None], int]"  [call-overload]
outlines/processors/tracking.py:246: note: Possible overload variants:
outlines/processors/tracking.py:246: note:     def __getitem__(self, SupportsIndex, /) -> Any
outlines/processors/tracking.py:246: note:     def __getitem__(self, slice[Any, Any, Any], /) -> list[Any]
outlines/processors/tracking.py:247: error: Value of type "dict[str, Any | Any | list[Any] | Any] | None" is not indexable  [index]
outlines/processors/tracking.py:247: error: No overload variant of "__getitem__" of "list" matches argument type "tuple[slice[None, None, None], int]"  [call-overload]
outlines/processors/tracking.py:247: note: Possible overload variants:
outlines/processors/tracking.py:247: note:     def __getitem__(self, SupportsIndex, /) -> Any
outlines/processors/tracking.py:247: note:     def __getitem__(self, slice[Any, Any, Any], /) -> list[Any]
outlines/processors/tracking.py:348: error: Library stubs not installed for "pandas"  [import-untyped]
outlines/processors/tracking.py:348: note: Hint: "python3 -m pip install pandas-stubs"
outlines/processors/tracking.py:348: note: (or run "mypy --install-types" to install all missing stub packages)
outlines/processors/tracking.py:348: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
outlines/processors/tracking.py:368: error: Item "list[Any]" of "Any | Any | list[Any] | Any" has no attribute "shape"  [union-attr]
outlines/processors/tracking.py:369: error: No overload variant of "__getitem__" of "list" matches argument type "tuple[slice[None, None, None], int]"  [call-overload]
outlines/processors/tracking.py:369: note: Possible overload variants:
outlines/processors/tracking.py:369: note:     def __getitem__(self, SupportsIndex, /) -> Any
outlines/processors/tracking.py:369: note:     def __getitem__(self, slice[Any, Any, Any], /) -> list[Any]
outlines/processors/tracking.py:370: error: No overload variant of "__getitem__" of "list" matches argument type "tuple[slice[None, None, None], int]"  [call-overload]
outlines/processors/tracking.py:370: note: Possible overload variants:
outlines/processors/tracking.py:370: note:     def __getitem__(self, SupportsIndex, /) -> Any
outlines/processors/tracking.py:370: note:     def __getitem__(self, slice[Any, Any, Any], /) -> list[Any]
outlines/processors/tracking.py:428: error: Item "None" of "OutlinesLogitsProcessor | None" has no attribute "tokenizer"  [union-attr]
outlines/processors/tracking.py:472: error: "SequenceGenerator" has no attribute "logits_processor"  [attr-defined]
outlines/processors/tracking.py:476: error: "SequenceGenerator" has no attribute "logits_processor"  [attr-defined]
outlines/processors/tracking.py:479: error: "SequenceGenerator" has no attribute "logits_processor"  [attr-defined]
outlines/processors/tracking.py:480: error: "SequenceGenerator" has no attribute "logits_processor"  [attr-defined]
outlines/processors/tracking.py:483: error: "SequenceGenerator" has no attribute "logits_processor"  [attr-defined]
tests/test_types.py:37: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
tests/test_types.py:93: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
tests/test_prompts.py:222: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
tests/test_prompts.py:223: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
tests/test_prompts.py:233: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
tests/test_prompts.py:234: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
tests/test_prompts.py:243: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
tests/test_prompts.py:244: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
tests/test_function.py:16: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
tests/processors/test_tracking.py:4: error: Library stubs not installed for "pandas"  [import-untyped]
```

**Comment by rlouf** [comment]
I can take a look.

**Comment by rlouf** [comment]
I fixed the formatting issues. I'll do a little refactoring and then we'll be good to merge.

**Comment by cpfiffer** [comment]
As an idle thought here -- is there an interface available to us where we could wrap the resulting generated object with the logits, rather than store it in the logit processor as I have here?

Currently we only return strings from `generator` calls, but is there an obvious + simple interface for providing a `Result(value=..., logits=...)` object? My sense is that this isn't likely to be simple, but the devex would probably be better.

Might be a "kick it down the road" thing, but curious if @rlouf @torymur @RobinPicard had an idea of whether this would be simple to do.

If it is simple, I could try refactoring this code to store logits in a response value.

**Comment by rlouf** [comment]
You'd need to create an object that stores the logits but otherwise behaves like a string. Not impossible.

**Comment by cpfiffer** [comment]
I am going to close this for now -- I'll revisit this later for v1.0 and generally tidy up the interface.
